api/src/main/java/org/openmrs/module/pihcore/event/PatientUpdateEventConsumer.java

```diff
@@ -0,0 +1,209 @@
+    public void startup() {
+        try {
+            statusDb = new Rocks(new File(config.getContext().getModuleDataDir(), "status.db"));
+            snapshotInitialized = BooleanUtils.isTrue(statusDb.get("snapshotInitialized"));
+        }
+        catch (Exception e) {
+            throw new RuntimeException("An error occurred starting up", e);
+        }
+    }
+
+    @Override
+    public void shutdown() {
+        statusDb.close();
+    }
+
+    @Override
+    public void accept(DbEvent event) {
+        if (!snapshotInitialized) {
+            performInitialSnapshot();
+            statusDb.put("snapshotInitialized", Boolean.TRUE);
+            snapshotInitialized = true;
+        }
+        Set<Integer> patientIds = getPatientIdsForEvent(event);
+        if (patientIds.isEmpty()) {
+            log.warn("Not handling event as not mapped to patient: " + event);
+        }
+        else {
+            for (Integer patientId : patientIds) {
+                handlePatientEvent(patientId, event);
+            }
+        }
+    }
+
+    /**
+     * For the given event, returns the associated patient ids
+     * @param event the event to retrieve the related patient ids
+     * @return the patient ids associated with the given event
+     */
+    public Set<Integer> getPatientIdsForEvent(DbEvent event) {
+        Set<Integer> ret = new HashSet<>();
+        for (String key : patientKeys.keySet()) {
+            String keyTable = patientKeys.get(key);
+            Integer value = event.getValues().getInteger(key);
+            if (value != null) {
+                SqlBuilder sql = new SqlBuilder();
+                sql.select("p.patient_id").from("patient", "p");
+                if (keyTable.equals("person")) {
+                    sql.where("p.patient_id = ?");
+                }
+                else {
+                    sql.innerJoin(keyTable, "x", "patient_id", "p", "patient_id");
+                    sql.where("x.").append(keyTable).append(" = ").append("?");
+                }
+                Integer patientId = database.executeQuery(sql.toString(), new ScalarHandler<>(1), value);
+                if (patientId == null) {
+                    if (!keyTable.equals("person")) {
+                        throw new RuntimeException("Unable to retrieve patient_id for: " + event);
+                    }
+                }
+                else {
+                    ret.add(patientId);
+                    if (!event.getTable().equals("relationship") || ret.size() == 2) {
+                        break;
+                    }
+                }
+            }
+        }
+        return ret;
+    }
+
+    /**
+     * Once a particular event is connected ot a particular patient, this tracks this change for the patient
+     * @param patientId the patient id related to the event
+     * @param event the event to track
+     */
+    public void handlePatientEvent(Integer patientId, DbEvent event) {
+        if (log.isDebugEnabled()) {
```

Comment:
For every event that we determine is an event for a particular patient, we update the db_event table with the timestamp of this event (based on the when the event occurred in the MySQL binlog), and also indicate whether the patient is deleted (based on either a hard delete of the row in the DB or based on a soft-delete if the patient is voided)